File: Python/0015-3Sum.py
# Time O(n^2)
# Space O(1)
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        result = []
        for i in range(len(nums)-2):
            if i > 0 and nums[i] == nums[i-1]:
                continue
            sum = nums[i]*-1
            low, high = i+1, len(nums)-1

            while low<high:
                if nums[low]+nums[high]==sum:
                    result.append([nums[i],nums[low],nums[high]])
                    while low<high and nums[low]==nums[low+1]:
                        low+=1
                    while low<high and nums[high]==nums[high-1]:
                        high-=1
                    low+=1
                    high-+1
                elif nums[low] + nums[high] > sum:
                    high-=1
                else:
                    low+=1
        return result

# Notes:
# If I fix two elements, there will only be one possible choice for third one.
# Here sorting is done so that we can avoid duplicate elements

File: C++/0015-3Sum.cpp
// Tags: Array Two-pointers Amazon
// Time O(n^2)
// Space O(1)
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> output;
        if(nums.size() <= 2) {
            return output;
        }
        sort(nums.begin(), nums.end());
        for(int i = 0; i < static_cast<int>(nums.size()) - 2; i++) {
            // or we can skip checking these conditions here and put a
            // while loop below as shown
            if (i == 0 || (i > 0 && nums[i] != nums[i-1])) {
                int low = i + 1;
                int high = nums.size() - 1;
                int sum = 0 - nums[i];
                while(low < high){
                    if (nums[low] + nums[high] == sum){
                        output.push_back({nums[i], nums[low], nums[high]});
                        while (low < high && nums[low] == nums[low+1])
                            low++;
                        while (low < high && nums[high] == nums[high-1])
                            high--;
                        low++;
                        high--;
                } else if (nums[low] + nums[high] > sum){
                        high--;
                    } else {
                        low++;
                    }
                }
            }
            // while (i < nums.size() - 2 && nums[i + 1] == nums[i])
            //  ++i;
        }
        return output;
    }
};

/*
 If I fix two elements, there will only be one possible choice for third one.
 Here sorting is done so that we can avoid duplicate elements
*/
